src/dex/ks-elastic/contract-math/ks-elastic-math.ts

```diff
@@ -0,0 +1,351 @@
+      invariant(sqrtPriceLimitX96 < poolState.sqrtPriceX96, 'RATIO_CURRENT');
+    } else {
+      invariant(sqrtPriceLimitX96 < TickMath.MAX_SQRT_RATIO, 'RATIO_MAX');
+      invariant(sqrtPriceLimitX96 > poolState.sqrtPriceX96, 'RATIO_CURRENT');
+    }
+    if (poolState.sqrtPriceX96 < 0n) {
+      console.log('PoolState====', poolState);
+    }
+    const exactInput = amountSpecified >= ZERO;
+    const state = {
+      amountSpecifiedRemaining: amountSpecified,
+      amountCalculated: ZERO,
+      baseL: poolState.liquidity,
+      reinvestL: poolState.reinvestLiquidity,
+      sqrtPriceX96: poolState.sqrtPriceX96,
+      tick: poolState.currentTick,
+    };
+
+    while (
+      state.amountSpecifiedRemaining != ZERO &&
+      state.sqrtPriceX96 != sqrtPriceLimitX96
+    ) {
+      const step = {
+        sqrtPriceStartX96: 0n,
+        tickNext: 0n,
+        initialized: false,
+        sqrtPriceNextX96: 0n,
+        amountIn: 0n,
+        amountOut: 0n,
+        feeAmount: 0n,
+        deltaL: 0n,
+      };
+      step.sqrtPriceStartX96 = state.sqrtPriceX96;
+      const [ticketNext, initialized] =
+        TickList.nextInitializedTickWithinFixedDistance(
+          tickList,
+          Number(state.tick),
+          zeroForOne,
+          480,
+        );
+
+      step.tickNext = BigInt(ticketNext);
+      step.initialized = initialized;
+
+      if (step.tickNext < TickMath.MIN_TICK) {
+        step.tickNext = TickMath.MIN_TICK;
+      } else if (step.tickNext > TickMath.MAX_TICK) {
+        step.tickNext = TickMath.MAX_TICK;
+      }
+
+      step.sqrtPriceNextX96 = TickMath.getSqrtRatioAtTick(step.tickNext);
+      [state.sqrtPriceX96, step.amountIn, step.amountOut, step.deltaL] =
+        SwapMath.computeSwapStepPromm(
+          state.sqrtPriceX96,
+          (
+            zeroForOne
+              ? step.sqrtPriceNextX96 < sqrtPriceLimitX96
+              : step.sqrtPriceNextX96 > sqrtPriceLimitX96
+          )
+            ? sqrtPriceLimitX96
+            : step.sqrtPriceNextX96,
+          state.baseL + state.reinvestL,
+          state.amountSpecifiedRemaining,
+          poolState.fee,
+          exactInput,
+          zeroForOne,
+        );
+
+      state.amountSpecifiedRemaining -= step.amountIn;
+      state.amountCalculated += step.amountOut;
+      state.reinvestL += step.deltaL;
+
+      if (state.sqrtPriceX96 !== step.sqrtPriceNextX96) {
+        state.tick = TickMath.getTickAtSqrtRatio(state.sqrtPriceX96);
+        break;
+      }
+      if (step.initialized) {
+        let tick = TickList.getTick(tickList, Number(step.tickNext));
+        let liquidityNet = tick.liquidityNet;
+
+        liquidityNet = zeroForOne ? -liquidityNet : liquidityNet;
+        state.baseL = LiquidityMath.addDelta(state.baseL, liquidityNet);
+      }
+      state.tick = zeroForOne ? step.tickNext - 1n : step.tickNext;
+    }
+    if (isSell)
+      return BigInt.asUintN(
```

Comment:
Don't ask me how, but I was able to resolve most of the issues reported in https://github.com/paraswap/paraswap-dex-lib/pull/227#issuecomment-1347558181 by applying this hack

```
return BigInt.asUintN(
  256,
  -(state.amountSpecifiedRemaining == 0n
    ? state.amountCalculated
    : state.amountSpecifiedRemaining)
);
```

Reply:
You can see how most of the stuff works fine

```
[2022-12-12T22:47:36.732] [DEBUG] dex - kyberswapElastic 0xae338fe2521c609437ab996f187819ab59082f94 inSync: {
  fetched_: { amount: '0.000000980069788427 WETH', duration: '501ms' },
  offChain: { amount: '0.000000980069791754 WETH', duration: '0ms' }
}
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:47:51.035] [DEBUG] dex - kyberswapElastic 0xae338fe2521c609437ab996f187819ab59082f94 inSync: {
  fetched_: { amount: '0.000029402971461697 WETH', duration: '381ms' },
  offChain: { amount: '0.000029402974651063 WETH', duration: '1ms' }
}
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:49:24.791] [INFO] pool - Initial state loaded for pool 0x114de2affc6a335433dbe9d3d51a8f31a5591fdf
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:49:25.536] [DEBUG] dex - kyberswapElastic 0x114de2affc6a335433dbe9d3d51a8f31a5591fdf inSync: {
  fetched_: { amount: '0.016396720057456019 WETH', duration: '366ms' },
  offChain: { amount: '0.016396720163999999 WETH', duration: '1ms' }
}
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:49:39.219] [DEBUG] dex - kyberswapElastic 0xae338fe2521c609437ab996f187819ab59082f94 inSync: {
  fetched_: { amount: '0.000098009957211371 WETH', duration: '422ms' },
  offChain: { amount: '0.000098009990481065 WETH', duration: '0ms' }
```

but there might still be instances of 0 results e.g.

```
[2022-12-12T22:51:35.278] [INFO] pool - Initial state loaded for pool 0x7acbea3b8ab7cdf4a595c6ed81e7d3e26038d494
node_modules/log4js/lib/appenders/console.js:6
[2022-12-12T22:51:36.306] [DEBUG] dex - kyberswapElastic 0x7acbea3b8ab7cdf4a595c6ed81e7d3e26038d494 inSync: {
  fetched_: { amount: '365.088392203183658678 wstETH', duration: '664ms' },
  offChain: { amount: '0.0 wstETH', duration: '15ms' }
}
```

Reply:
can you share how to debug like you? It is the main key for me to solving all current problems.

Reply:
hey @linhbkhn95, I do calculate the off chain result via this snippet

```ts
export async function fetchQuote(
  tokenIn: string,
  tokenOut: string,
  amountIn: string,
  pool: string
): Promise<BigNumber> {
  const response = await axios.get<Quote>(
    'https://aggregator-api.kyberswap.com/arbitrum/route/encode',
    {
      params: {
        tokenIn: tokenIn,
        tokenOut: tokenOut,
        amountIn: amountIn,
        dexes: 'kyberswapv2',
        saveGas: '1',
        to: '0x0000000000000000000000000000000000000000',
      },
    }
  );

  const quote = response.data;

  // We need to find the swap that does only use the Pool that we are trying to get a quote for
  const poolSwap = quote.swaps.find((swaps) => {
    return swaps.length === 1 && isSameAddress(pool, swaps[0].pool);
  });

  if (poolSwap === undefined) {
    throw new Error('Unable to find quote for pool: ' + pool);
  }

  const { amountOut } = poolSwap[0];
  return BigNumber.from(amountOut);
}
```

Reply:
can you provide the full code or ref from where?

Reply:
hey @linhbkhn95 I do run `fetchQuote` (my implementation above) and `getOutputAmountProMM` (your implementation) whenever a swap happen on a pool, then print out the two amounts. I am happy to provide the full snippet on Discord, my username is `adedade#3981` add me :)

Reply:
Thanks, I sent my invitation to add friend